src/gmres_e.m

```diff
@@ -0,0 +1,426 @@
+    xm = xInitial + V * minimizer;
+
+    % Update residual norm, iterations, and relative residual vector
+    res(restart + 1, :) = abs(g(m + 1, 1));
+    iter(restart + 1, :) = restart + 1;
+    relresvec(restart + 1, :) = res(restart + 1, :) / res(1, 1);
+
+    % Check convergence
+    if relresvec(restart + 1, :) < tol %|| size(relresvec, 1) == maxit
+        % We reached convergence.
+        flag = 1;
+        x = xm;
+        time = toc();
+        return
+    else
+        % We have not reached convergence.
+        % For GMRES-E, an eigenvalue problem is solved.
+        % Eigenvalue problem setup, from [1], p. 1161, step 5
+        Fold = H(1:s, 1:s)';
+        G = Rs' * Rs;
+        opts.tol = tol;
+        opts.v0 = ones(size(Fold, 1), 1);
+        E = zeros(s, k);
+        D = zeros(k, 1);
+
+        % Compute the harmonic Ritz vectors
+        [E2, D2] = eigs(Fold, G, k, 'LM', opts);
+        for p = 1:k
+            D(p, 1) = abs(D2(p, p));
+        end
+        [~, I] = sort(D, 1);
+        for q = 1:k
+            E(:, q) = E2(:, I(q, 1));
+        end
+        dy0 = V * E; % yi = Q * gi???
+
+        dy = zeros(n, 1);
+
+        % If dy0 has complex components, we separate each eigenvector
+        % into real and complex parts and treat as two distinct vectors
+        if isreal(dy0) == 0
+            ij = 1;
+            jj = 0;
+            while size(dy, 2) <= k && ij <= k
+                if isreal(dy0(:, ij)) == 0 && norm(real(dy0(:, ij))) > 0
+                    dy(:, jj + 1) = real(dy0(:, ij));
+                    jj = size(dy, 2);
+                    if ij <= k
+                        dy(:, jj + 1) = abs(imag(dy0(:, ij)) * sqrt(1));
+                        jj = size(dy, 2);
+                        if ij < k
+                            if dy0(:, ij) == conj(dy0(:, ij + 1))
+                                ij = ij + 2;
+                            else
+                                ij = ij + 1;
+                            end
+                        end
+                    end
+                else
+                    dy(:, jj + 1) = dy0(:, ij);
+                    ij = ij + 1;
+                    jj = size(dy, 2);
+                end
+            end
+        else
+            dy = dy0;
+        end
+
+        % Update and restart.
+        restart = restart + 1;
+    end
+
+    % Empty matrix E for next outer iteration
+    E = zeros(s, k);
+
+    % ---> GMRES-E Algorithm for restart > 1 or
+    % GMRES-E(m, k)
+
+    while flag == 0 && restart <= maxit
+
+        % Compute normalized residual vector
+        r = b - A * xm;
+        beta = norm(r);
+        v1 = r / beta;
+
+        % Augmented Gram-Schmidt Arnoldi iteration
+        % Notice that for GMRES-E, we need E,
+        % a n-by-k matrix, the matrix of eigenvectors
+        % from the last outer iteration.
+        [H, V, s] = ...
+            augmented_gram_schmidt_arnoldi(A, v1, m, fliplr(dy(:, 1:k)));
+        % Patch: to be solved
```

Comment:
Please resolve this or create an issue of this question will remain unresolved.

Reply:
I will move for now to a new issue.